src/raft/raft.rs

```diff
@@ -1723,24 +1836,32 @@ impl<T: Storage> Raft<T> {
         );
 
         let nodes = meta.get_conf_state().get_nodes();
-        let prs = FlatMap::with_capacity(meta.get_conf_state().get_nodes().len());
-        self.prs = Some(prs);
+        let learners = meta.get_conf_state().get_learners();
+        self.prs = Some(ProgressSet::new(nodes.len(), learners.len()));
 
-        let self_id = self.id;
-        let last_index = self.raft_log.last_index();
-        for &n in nodes {
-            let next_index = last_index + 1;
-            let matched = if n == self_id { next_index - 1 } else { 0 };
-            self.set_progress(n, matched, next_index);
-            info!(
-                "{} restored progress of {} [{:?}]",
-                self.tag,
-                n,
-                self.get_prs().get(&n)
-            );
+        {
+            let mut restore_nodes = |nodes: &[u64], is_learner: bool| for &n in nodes {
```

Comment:
You can use `for &(is_learner, nodes) in &[(false, nodes), (true, learners)] {xxx}`, so brackets are not necessary.